File: android/src/main/kotlin/com/encorekit/encore_flutter_sdk/EncoreFlutterPlugin.kt
package com.encorekit.encore_flutter_sdk

import android.app.Activity
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import com.encorekit.encore.Encore
import com.encorekit.encore.core.canonical.user.UserAttributes
import com.encorekit.encore.core.infrastructure.logging.LogLevel
import com.encorekit.encore.features.offers.PresentationResult
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

class EncoreFlutterPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {

    private lateinit var channel: MethodChannel
    private var activity: Activity? = null
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    // -- FlutterPlugin --

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "com.encorekit/encore")
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        scope.cancel()
    }

    // -- ActivityAware --

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    // -- MethodCallHandler --

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "configure" -> handleConfigure(call, result)
            "identify" -> handleIdentify(call, result)
            "setUserAttributes" -> handleSetUserAttributes(call, result)
            "reset" -> handleReset(result)
            "showPlacement" -> handleShowPlacement(call, result)
            "registerOnPurchaseRequest" -> registerOnPurchaseRequest(result)
            "registerOnPurchaseComplete" -> registerOnPurchaseComplete(result)
            "registerOnPassthrough" -> registerOnPassthrough(result)
            else -> result.notImplemented()
        }
    }

    // -- Configure --

    private fun handleConfigure(call: MethodCall, result: Result) {
        val apiKey = call.argument<String>("apiKey")
        if (apiKey == null) {
            result.error("INVALID_ARGS", "apiKey is required", null)
            return
        }

        val currentActivity = activity
        if (currentActivity == null) {
            result.error("NO_ACTIVITY", "No activity available for configure", null)
            return
        }

        val logLevel = parseLogLevel(call.argument<String>("logLevel"))

        Encore.shared.configure(currentActivity.applicationContext, apiKey, logLevel)
        result.success(null)
    }

    // -- Identity --

    private fun handleIdentify(call: MethodCall, result: Result) {
        val userId = call.argument<String>("userId")
        if (userId == null) {
            result.error("INVALID_ARGS", "userId is required", null)
            return
        }

        val attributes = parseUserAttributes(call.argument<Map<String, Any>>("attributes"))
        Encore.shared.identify(userId, attributes)
        result.success(null)
    }

    private fun handleSetUserAttributes(call: MethodCall, result: Result) {
        val attrsMap = call.argument<Map<String, Any>>("attributes")
        if (attrsMap == null) {
            result.error("INVALID_ARGS", "attributes is required", null)
            return
        }

        val attributes = parseUserAttributes(attrsMap) ?: UserAttributes()
        Encore.shared.setUserAttributes(attributes)
        result.success(null)
    }

    // -- Reset --

    private fun handleReset(result: Result) {
        Encore.shared.reset()
        result.success(null)
    }

    // -- Placements --

    private fun handleShowPlacement(call: MethodCall, result: Result) {
        val placementId = call.argument<String>("placementId")
        val currentActivity = activity

        if (currentActivity == null) {
            result.error("NO_ACTIVITY", "No activity available for showPlacement", null)
            return
        }

        scope.launch {
            try {
                val nativeResult = Encore.shared.placement(placementId).show(currentActivity)
                result.success(serializePresentationResult(nativeResult))
            } catch (e: Exception) {
                result.error("SHOW_FAILED", e.message, null)
            }
        }
    }

    // -- On-Demand Handler Registration --

    private fun registerOnPurchaseRequest(result: Result) {
        Encore.shared.onPurchaseRequest { purchaseRequest ->
            suspendCancellableCoroutine { continuation ->
                val fields = purchaseRequest.javaClass.declaredFields
                    .filter { it.type == String::class.java }
                    .onEach { it.isAccessible = true }
                val productId = fields.getOrNull(0)?.get(purchaseRequest) as? String ?: ""
                val placementId = fields.getOrNull(1)?.get(purchaseRequest) as? String ?: ""
                scope.launch {
                    channel.invokeMethod("onPurchaseRequest", mapOf(
                        "productId" to productId,
                        "placementId" to placementId,
                    ), object : MethodChannel.Result {
                        override fun success(result: Any?) {
                            continuation.resume(Unit)
                        }
                        override fun error(code: String, message: String?, details: Any?) {
                            continuation.resumeWithException(
                                RuntimeException(message ?: "Purchase failed")
                            )
                        }
                        override fun notImplemented() {
                            continuation.resume(Unit)
                        }
                    })
                }
            }
        }
        result.success(null)
    }

    private fun registerOnPurchaseComplete(result: Result) {
        Encore.shared.onPurchaseComplete { billingResult, productId ->
            scope.launch {
                channel.invokeMethod("onPurchaseComplete", mapOf(
                    "result" to mapOf(
                        "productId" to billingResult.productId,
                        "purchaseToken" to billingResult.purchaseToken,
                        "orderId" to billingResult.orderId,
                    ),
                    "productId" to productId,
                ))
            }
        }
        result.success(null)
    }

    private fun registerOnPassthrough(result: Result) {
        Encore.shared.onPassthrough { placementId ->
            scope.launch {
                channel.invokeMethod("onPassthrough", mapOf(
                    "placementId" to placementId,
                ))
            }
        }
        result.success(null)
    }

    // -- Serialization Helpers --

    private fun serializePresentationResult(nativeResult: PresentationResult): Map<String, Any?> {
        return when (nativeResult) {
            is PresentationResult.Completed -> mapOf(
                "type" to "granted",
                "offerId" to nativeResult.offerId,
                "campaignId" to nativeResult.campaignId,
            )
            is PresentationResult.Dismissed -> mapOf(
                "type" to "notGranted",
                "reason" to nativeResult.reason.value,
            )
            is PresentationResult.NoOffers -> mapOf(
                "type" to "notGranted",
                "reason" to "no_offer_available",
            )
        }
    }

    private fun parseLogLevel(value: String?): LogLevel {
        return when (value?.uppercase()) {
            "DEBUG" -> LogLevel.DEBUG
            "INFO" -> LogLevel.INFO
            "WARN" -> LogLevel.WARN
            "ERROR" -> LogLevel.ERROR
            else -> LogLevel.NONE
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun parseUserAttributes(map: Map<String, Any>?): UserAttributes? {
        if (map == null || map.isEmpty()) return null

        val custom = (map["custom"] as? Map<String, String>) ?: emptyMap()

        return UserAttributes(
            email = map["email"] as? String,
            firstName = map["firstName"] as? String,
            lastName = map["lastName"] as? String,
            phoneNumber = map["phoneNumber"] as? String,
            postalCode = map["postalCode"] as? String,
            city = map["city"] as? String,
            state = map["state"] as? String,
            countryCode = map["countryCode"] as? String,
            latitude = map["latitude"] as? String,
            longitude = map["longitude"] as? String,
            dateOfBirth = map["dateOfBirth"] as? String,
            gender = map["gender"] as? String,
            language = map["language"] as? String,
            subscriptionTier = map["subscriptionTier"] as? String,
            monthsSubscribed = map["monthsSubscribed"] as? String,
            billingCycle = map["billingCycle"] as? String,
            lastPaymentAmount = map["lastPaymentAmount"] as? String,
            lastActiveDate = map["lastActiveDate"] as? String,
            totalSessions = map["totalSessions"] as? String,
            custom = custom,
        )
    }
}

File: ios/Classes/EncoreFlutterPlugin.swift
import Flutter
import UIKit
import Encore

public class EncoreFlutterPlugin: NSObject, FlutterPlugin {
    private var channel: FlutterMethodChannel

    init(channel: FlutterMethodChannel) {
        self.channel = channel
        super.init()
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "com.encorekit/encore",
            binaryMessenger: registrar.messenger()
        )
        let instance = EncoreFlutterPlugin(channel: channel)
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    // MARK: - MethodChannel Handler

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any]

        switch call.method {
        case "configure":
            handleConfigure(args: args, result: result)
        case "identify":
            handleIdentify(args: args, result: result)
        case "setUserAttributes":
            handleSetUserAttributes(args: args, result: result)
        case "reset":
            handleReset(result: result)
        case "showPlacement":
            handleShowPlacement(args: args, result: result)
        case "registerOnPurchaseRequest":
            registerOnPurchaseRequest(result: result)
        case "registerOnPurchaseComplete":
            registerOnPurchaseComplete(result: result)
        case "registerOnPassthrough":
            registerOnPassthrough(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Configure

    private func handleConfigure(args: [String: Any]?, result: @escaping FlutterResult) {
        guard let apiKey = args?["apiKey"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "apiKey is required", details: nil))
            return
        }

        let logLevel = parseLogLevel(args?["logLevel"] as? String)

        Encore.shared.configure(apiKey: apiKey, logLevel: logLevel)
        result(nil)
    }

    // MARK: - Identity

    private func handleIdentify(args: [String: Any]?, result: @escaping FlutterResult) {
        guard let userId = args?["userId"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "userId is required", details: nil))
            return
        }

        let attributes = parseUserAttributes(args?["attributes"] as? [String: Any])
        Encore.shared.identify(userId: userId, attributes: attributes)
        result(nil)
    }

    private func handleSetUserAttributes(args: [String: Any]?, result: @escaping FlutterResult) {
        guard let attrsMap = args?["attributes"] as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGS", message: "attributes is required", details: nil))
            return
        }

        let attributes = parseUserAttributes(attrsMap)
        Encore.shared.setUserAttributes(attributes ?? UserAttributes())
        result(nil)
    }

    // MARK: - Reset

    private func handleReset(result: @escaping FlutterResult) {
        Encore.shared.reset()
        result(nil)
    }

    // MARK: - Placements

    private func handleShowPlacement(args: [String: Any]?, result: @escaping FlutterResult) {
        let placementId = args?["placementId"] as? String

        Task { @MainActor in
            do {
                let nativeResult = try await Encore.shared.placement(placementId).show()
                result(self.serializePresentationResult(nativeResult))
            } catch {
                result(FlutterError(
                    code: "SHOW_FAILED",
                    message: error.localizedDescription,
                    details: nil
                ))
            }
        }
    }

    // MARK: - On-Demand Handler Registration

    private func registerOnPurchaseRequest(result: @escaping FlutterResult) {
        Encore.shared.onPurchaseRequest { [weak self] productId, placementId in
            try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
                DispatchQueue.main.async {
                    self?.channel.invokeMethod("onPurchaseRequest", arguments: [
                        "productId": productId,
                        "placementId": placementId ?? "",
                    ]) { response in
                        if let error = response as? FlutterError {
                            continuation.resume(throwing: NSError(
                                domain: "EncoreFlutter",
                                code: -1,
                                userInfo: [NSLocalizedDescriptionKey: error.message ?? "Purchase failed"]
                            ))
                        } else {
                            continuation.resume()
                        }
                    }
                }
            }
        }
        result(nil)
    }

    private func registerOnPurchaseComplete(result: @escaping FlutterResult) {
        Encore.shared.onPurchaseComplete { [weak self] transaction, productId in
            DispatchQueue.main.async {
                self?.channel.invokeMethod("onPurchaseComplete", arguments: [
                    "result": [
                        "productId": productId,
                        "transactionId": String(transaction.id),
                    ],
                    "productId": productId,
                ])
            }
        }
        result(nil)
    }

    private func registerOnPassthrough(result: @escaping FlutterResult) {
        Encore.shared.onPassthrough { [weak self] placementId in
            DispatchQueue.main.async {
                self?.channel.invokeMethod("onPassthrough", arguments: [
                    "placementId": placementId ?? "",
                ])
            }
        }
        result(nil)
    }

    // MARK: - Serialization Helpers

    private func serializePresentationResult(_ nativeResult: PresentationResult) -> [String: Any] {
        switch nativeResult {
        case .granted:
            return ["type": "granted"]
        case .notGranted(let reason):
            return ["type": "notGranted", "reason": reason.rawValue]
        @unknown default:
            return ["type": "notGranted", "reason": "unknown"]
        }
    }

    private func parseLogLevel(_ value: String?) -> Encore.LogLevel {
        switch value?.uppercased() {
        case "DEBUG": return .debug
        case "INFO": return .info
        case "WARN": return .warn
        case "ERROR": return .error
        default: return .none
        }
    }

    private func parseUserAttributes(_ map: [String: Any]?) -> UserAttributes? {
        guard let map = map, !map.isEmpty else { return nil }

        var customDict: [String: String] = [:]
        if let custom = map["custom"] as? [String: String] {
            customDict = custom
        }

        return UserAttributes(
            email: map["email"] as? String,
            firstName: map["firstName"] as? String,
            lastName: map["lastName"] as? String,
            phoneNumber: map["phoneNumber"] as? String,
            postalCode: map["postalCode"] as? String,
            city: map["city"] as? String,
            state: map["state"] as? String,
            countryCode: map["countryCode"] as? String,
            latitude: map["latitude"] as? String,
            longitude: map["longitude"] as? String,
            dateOfBirth: map["dateOfBirth"] as? String,
            gender: map["gender"] as? String,
            language: map["language"] as? String,
            subscriptionTier: map["subscriptionTier"] as? String,
            monthsSubscribed: map["monthsSubscribed"] as? String,
            billingCycle: map["billingCycle"] as? String,
            lastPaymentAmount: map["lastPaymentAmount"] as? String,
            lastActiveDate: map["lastActiveDate"] as? String,
            totalSessions: map["totalSessions"] as? String,
            custom: customDict
        )
    }
}
